packages/telemetry/src/context-aware-slog.js

```diff
@@ -0,0 +1,475 @@
+        .run(),
+  );
+
+  return {
+    /**
+     *
+     * @param {Context} context
+     */
+    persistContext: context => {
+      /** @type {Array<string>} */
+      const keys = [];
+      const values = [];
+
+      for (const [key, value] of Object.entries(context)) {
+        keys.push(`[${key}]`);
+        if (value === undefined) values.push(null);
+        else
+          values.push(
+            DATABASE_SCHEMA[key].startsWith(DATABASE_TYPES.TEXT)
+              ? `'${value}'`
+              : value,
+          );
+      }
+
+      databaseInstance
+        .prepare(
+          `INSERT OR REPLACE INTO '${DATABASE_TABLE_NAME}' (key, ${keys.join(', ')})
+           VALUES ('${DATABASE_TRIGGER_CONTEXT_KEY_NAME}', ${values.join(', ')})`,
+        )
+        .run();
+    },
+
+    restoreContext: () => {
+      /**
+       * @type {Context & {key?: string} | undefined}
+       */
+      // @ts-expect-error
+      const row = databaseInstance
+        .prepare(
+          `SELECT * FROM '${DATABASE_TABLE_NAME}' WHERE key = '${DATABASE_TRIGGER_CONTEXT_KEY_NAME}'`,
+        )
+        .get();
+
+      delete row?.key;
+
+      return row || null;
+    },
+  };
+};
+
+const stringify = data =>
+  JSON.stringify(data, (_, value) =>
+    typeof value === 'bigint' ? Number(value) : value,
+  );
+
+/**
+ *
+ * @param {{env: typeof process.env}} options
+ */
+export const makeSlogSender = async options => {
+  const { CHAIN_ID, OTEL_EXPORTER_OTLP_ENDPOINT } = options.env;
+  if (!OTEL_EXPORTER_OTLP_ENDPOINT)
+    return console.warn(
+      'Ignoring invocation of slogger "context-aware-slog" without the presence of "OTEL_EXPORTER_OTLP_ENDPOINT" envrionment variable',
+    );
+
+  const loggerProvider = new LoggerProvider({
+    resource: new Resource(getResourceAttributes(options)),
+  });
+  loggerProvider.addLogRecordProcessor(
+    new SimpleLogRecordProcessor(new OTLPLogExporter({ keepAlive: true })),
+  );
+
+  logs.setGlobalLoggerProvider(loggerProvider);
+  const logger = logs.getLogger('default');
+
+  const db = await getDatabaseInstance();
+
+  /** @type Array<Context | null> */
+  let [
+    blockContext,
+    crankContext,
+    initContext,
+    lastPersistedTriggerContext,
+    replayContext,
+    triggerContext,
+  ] = [null, null, null, null, null, null];
+
+  /**
+   * @param {Context} context
+   */
+  const persistContext = context => {
+    lastPersistedTriggerContext = context;
+    db.persistContext(context);
+  };
+
+  const restoreContext = () => {
+    if (lastPersistedTriggerContext) return lastPersistedTriggerContext;
+    return db.restoreContext();
+  };
+
+  /**
+   * @param {Slog} slog
+   */
+  const slogSender = ({ monotime, time: timestamp, ...body }) => {
+    const finalBody = { ...body };
+
+    /** @type {LogAttributes} */
+    let logAttributes = {
+      'chain-id': String(CHAIN_ID),
+      'process.uptime': monotime,
+      timestamp,
```

Comment:
I think there could be problems with timezone if we tinker with timestamp field. I suggest letting the timestamp be populated automatically by the sink and use the `observedTimestamp` field

Reply:
Observed timestamp should be filed automatically by the SDK, no?

What is the concern with tz? I believe the timestamp comes in as an iso string datetime.

Reply:
The issue is with the sink (google cloud logging in this case). Due to timezone difference, the reported timestamp can be in future for the sink which causes failures. I think timestamp should be populated by the sink itself.

Reply:
That really doesn't make sense to me. Afaik the [slog event generator uses](https://github.com/Agoric/agoric-sdk/blob/df2faae7e5aa4bba657962f3b33aaeb7a3953c91/packages/SwingSet/src/controller/controller.js#L172-L173) the POSIX time for this value, and it does so on the same machine, so there is no reason that the timestamp generated by the otel SDK and the timestamp we generate would be having such a timezones difference. I would really like to be able to rely on our source of truth for time, so that we can keep our ability to backfill data by re-streaming captured slogs if necessary.

Reply:
I have tried it and google cloud sink gives error that the timestamp can’t be in future. But their limits page suggests that timestamp can be max 1 day in future so we might be able to fix it. Even if we could, the same page also suggest that maximum 30 days in the past are allowed so we won’t be able to report all the historic data using the timestamp field:
https://cloud.google.com/logging/quotas#log-limits
I would say we don’t use timestamp field and use either a label or a jsonPayload field for our time. We would still be able to use them to filter logs for a time range.